modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -338,6 +348,31 @@ object ConsensusStateUpdater {
     private def pickMajority[A: Order](proposals: List[A]): Option[A] =
       proposals.foldMap(a => Map(a -> 1)).toList.map(_.swap).maximumOption.map(_._2)
 
+    private def pickValidatedMajority(
+      lastSignedArtifact: Signed[Artifact],
+      trigger: ConsensusTrigger,
+      resources: ConsensusResources[Artifact]
+    )(proposals: List[Hash]): F[Option[Hash]] = {
+      val sortedProposals = proposals.foldMap(a => Map(a -> 1)).toList.map(_.swap).sorted.reverse
+
+      def go(majorityHash: (Int, Hash), tail: List[(Int, Hash)]): F[Option[Hash]] =
+        resources.artifacts
+          .get(majorityHash._2)
+          .traverse { artifact =>
+            consensusFns.validateArtifact(lastSignedArtifact, trigger)(artifact)
+          }
+          .flatMap {
+            case None => none[Hash].pure[F]
+            case Some(Left(cause)) =>
+              logger.warn(cause)(s"Found invalid majority hash=${majorityHash._2.show} with occurrences=${majorityHash._1}") >>
+                tail.headOption.fold(none[Hash].pure[F])(go(_, tail.tail))
+            case Some(Right(_)) => majorityHash._2.some.pure[F]
+          }
+
+      sortedProposals.headOption
+        .fold(none[Hash].pure[F])(go(_, sortedProposals.tail))
```

Comment:
This line and 368 look exactly the same. You could get rid of this small duplication (along with `._1` and `._2`)  If you would redefine `go` to:

```scala
def go(proposals: List[(Int, Hash)]): F[Option[Hash]] =
  proposals match {
    case (occurrences, majorityHash) :: tail =>
      resources.artifacts
        .get(majorityHash)
        .traverse { artifact =>
          consensusFns.validateArtifact(lastSignedArtifact, trigger)(artifact)
        }
        .flatMap { maybeValidatedArtifact =>
          maybeValidatedArtifact.flatTraverse { artifactOrError =>
            artifactOrError.fold(
              cause =>
                logger.warn(cause)(s"Found invalid majority hash=${majorityHash.show} with occurrences=$occurrences") >> go(tail),
              _ => majorityHash.some.pure[F]
            )
          }
        }
    case Nil => none[Hash].pure[F]
  }
```